### PR TITLE
GUNDI-4998: eBird filtering logic & unit testing

### PR DESCRIPTION
**Pull request by vgarcia13**
### Related Ticket

https://allenai.atlassian.net/browse/GUNDI-4998

---

This pull request introduces improvements to the eBird event pulling workflow, focusing on more reliable event deduplication, dynamic lookback calculation, and timezone handling. It also adds new tests to verify these changes. 

The most important updates are grouped below.

**Connector now runs using a hourly `crontab_schedule`**

**Event Deduplication and Filtering**

* Added a new function `filter_ebird_events` to filter out eBird events that have already been processed, based on the latest observation datetime stored in state. This prevents duplicate submissions to Gundi.
* Updated the event submission logic in `action_pull_events` to use `filter_ebird_events` before sending data to Gundi, and to record the latest observation datetime after successful submission.

**Dynamic Lookback Calculation**

* Modified `action_pull_events` to dynamically calculate the number of days to look back for new observations, based on the last execution time stored in state. This ensures only new data since the last run is fetched. [[1]](diffhunk://#diff-f575ea29ea1c9649ccb87f9c38a0fd9a78b9f210c9e5e9d800e328ee30d08113R122-R152) [[2]](diffhunk://#diff-f575ea29ea1c9649ccb87f9c38a0fd9a78b9f210c9e5e9d800e328ee30d08113L117-R163)

**Timezone Handling**

* Updated `_transform_ebird_to_gundi_event` so that the `recorded_at` timestamp is always timezone-aware (UTC), improving consistency in event data.

**Testing Improvements**

* Added new tests in `test_handlers.py` to verify event filtering and transformation, including edge cases for timezone-aware and naive datetimes.

---

> [!NOTE]
> Make eBird pull stateful with timezone-aware deduplication, dynamic lookback, hourly scheduling, and tests.
> 
> - **Actions (eBird)**:
>   - Introduce `State` with `latest_observation_at`; load/persist state via `IntegrationStateManager` in `action_pull_events`.
>   - Compute dynamic lookback days from last observation; fetch observations accordingly and filter to only those newer than state.
>   - Persist max observed timestamp after successful send; return extracted count.
>   - Make datetimes timezone-aware (UTC) and serialize `recorded_at` via `.isoformat()`; add validators for `obsDt` and state.
>   - Schedule `action_pull_events` hourly with `@crontab_schedule`.
> - **Tests**:
>   - Add tests for event transformation and timezone handling in `app/actions/tests/test_handlers.py`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 8d8af94a656759b9b281fd0b6d50d28898db0724. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: app/actions/handlers.py [hunk 1]
```diff
@@ -87,6 +92,24 @@ def get_auth_config(integration):
...
     return AuthenticateConfig.parse_obj(auth_config.data)
 
+async def filter_ebird_events(integration_id: str, events: List[dict]) -> List[dict]:
+    saved_state = await state_manager.get_state(
+        integration_id,
+        "pull_events",
+        "latest_observation_datetime"
+    )
+    if saved_state:
+        latest_observation_datetime = datetime.fromisoformat(saved_state.get("latest_observation_datetime"))
```
[the remark this answers is not in the document]
Comment (chrisdoehring) [reply app/actions/handlers.py]: @vgarcia13 did you consider using a pydantic model for the state object? Often times that can help to encapsulate the validation, defaults, etc. rather than manually parsing data like this.

File: app/actions/handlers.py [hunk 2]
```diff
@@ -96,14 +119,37 @@ async def action_pull_events(integration:Integration, action_config: PullEventsC
...
+        parsed_latest_execution_time = datetime.fromtimestamp(latest_execution_time, tz=timezone.utc)
+
+        logger.info(f"Latest execution time found in state: {parsed_latest_execution_time.isoformat()} for integration ID: {str(integration.id)}")
+
+        # If it exists, adjust num_days to cover from that time to now
+        now = datetime.now(tz=timezone.utc)
+        delta = now - parsed_latest_execution_time
+        days_difference = math.ceil(delta.total_seconds() / SECONDS_IN_DAY)
+        lookback_days_to_fetch = days_difference
+        logger.info(f"Adjusted num_days to {lookback_days_to_fetch} to cover from latest execution time to now for integration ID: {str(integration.id)}")
```
[the remark this answers is not in the document]
Comment (chrisdoehring) [reply app/actions/handlers.py]: Using a model for state like suggested earlier could mitigate this kind of failure case.

File: app/actions/handlers.py [hunk 3]
```diff
@@ -124,19 +170,42 @@ async def action_pull_events(integration:Integration, action_config: PullEventsC
...
+            )
+            latest_time = max(filtered_events_to_send, key=lambda obs: obs["recorded_at"])["recorded_at"]
+            state = {"latest_observation_datetime": latest_time.isoformat()}
+            await state_manager.set_state(
+                str(integration.id),
+                "pull_events",
+                state,
+                "latest_observation_datetime"
+            )
+            events_extracted += len(filtered_events_to_send)
```
[the remark this answers is not in the document]
Comment (chrisdoehring) [reply app/actions/handlers.py]: @vgarcia13 please consider this suggestion.

**Review by chrisdoehring — CHANGES_REQUESTED** [review]
Thank you @vgarcia13 
I added some comments inline with Copilot's suggestions. Please have a look.

File: app/actions/handlers.py [hunk 4]
```diff
@@ -15,6 +17,14 @@
 state_manager = IntegrationStateManager()
 
 EBIRD_API = "https://api.ebird.org/v2"
+SECONDS_IN_DAY = 86400
+
+
+class State(BaseModel):
+    latest_observation_datetime: Optional[datetime]
+    latest_execution_time: Optional[datetime]
```
Comment (chrisdoehring) [inline app/actions/handlers.py]: It looks like you're storing separate state objects for these two different values. I suggest you create two different state models for those, so this won't be confusing the next time someone reads this.

File: app/actions/handlers.py [hunk 5]
```diff
@@ -96,14 +125,37 @@ async def action_pull_events(integration:Integration, action_config: PullEventsC
...
+    saved_latest_execution_time = await state_manager.get_state(
+        str(integration.id),
+        "pull_events",
+        "latest_execution_time"
+    )
+    if saved_latest_execution_time:
+        state = State.parse_obj(saved_latest_execution_time)
+        latest_execution_time = state.latest_execution_time
+
+        logger.info(f"Latest execution time found in state: {latest_execution_time.isoformat()} for integration ID: {str(integration.id)}")
```
Comment (chrisdoehring) [inline app/actions/handlers.py]: Since the State model includes this as Optional, it's possible it's null here, and would cause an error on call to isoformat() and any subsequent calls.

**Review by chrisdoehring — COMMENTED** [review]
That's an improvement, but you'll want to use validation and sensible defaults in State to really benefit. Copilot made some more comments too.

File: app/actions/handlers.py (hunk 4 above)
Comment (vgarcia13) [reply app/actions/handlers.py]: I created 2 different state models for better readability. I can't think in more validations there because I control which data is saved in the state and then parsed by these models, so I am sure tz info is set and the format is correct

File: app/actions/handlers.py [hunk 6]
```diff
@@ -1,20 +1,47 @@
...
 
 EBIRD_API = "https://api.ebird.org/v2"
+SECONDS_IN_DAY = 86400 # 24 hours * 60 minutes * 60 seconds
+
+
+class LatestObservationDatetimeState(BaseModel):
+    latest_observation_datetime: datetime
+
+    @validator('latest_observation_datetime', pre=True, always=True)
+    def clean_latest_observation_datetime(cls, v):
```
Comment (chrisdoehring) [inline app/actions/handlers.py]: This is still vulnerable to a value that is missing or is not a timestamp. In addition to following the copilot recommendation above, please consider that this validator is meant to guarantee that the resulting object has a valid value for `latest_obseravtion_datetime`.  So in fact your latest PR had this field as `Optional` and that is probably a good idea. With that, you get a default value of `None` which might be the best default value depending this model's use.

The validator can concentrate on the value you will actually get in the parsed data.

**Review by chrisdoehring — CHANGES_REQUESTED** [review]
Thanks Victor. I'm afraid there are still some changes necessary to properly handle the state objects. Have a look at the comments inline.

File: app/actions/handlers.py [hunk 7]
```diff
@@ -15,6 +17,57 @@
...
+            # Assuming v is a string
+            try:
+                parsed = datetime.fromisoformat(v)
+            except Exception:
+                # string is not in ISO format, return None
+                return None
+            else:
+                if parsed.tzinfo is None:
+                    parsed = parsed.replace(tzinfo=timezone.utc)
+                return parsed
```
Comment (chrisdoehring) [inline app/actions/handlers.py]: This got quite complicated. What i recommend is having a simple validator like this, to ensure the value is timezone aware.

```suggestion
class LatestObservationDatetimeState(BaseModel):
    latest_observation_datetime: Optional[datetime]
    @validator('latest_observation_datetime')
    def clean_latest_observation_datetime(cls, v):
        if isinstance(v, datetime):
            if v.tzinfo is None:
                return v.replace(tzinfo=timezone.utc)
        return v
```

Then when parsing data, catch validation errors just to catch unexpected cases (ex. value is a string that's not a validate date and time, or is over the wrong type).

**Review by chrisdoehring — CHANGES_REQUESTED** [review]
I would still prefer to have a simpler state model. I added a suggested edit to demonstrate.

File: app/actions/handlers.py (hunk 7 above)
Comment (vgarcia13) [reply app/actions/handlers.py]: I had something like this in this commit:

https://github.com/PADAS/gundi-integration-ebird/pull/20/commits/27a43acd4c18e50951159d5054846a27bed81da5

So I will return to that

**Review by chrisdoehring — APPROVED** [review]
